mp/src/game/server/momentum/mom_triggers.cpp

```diff
@@ -1578,3 +1579,345 @@ bool CNoGrenadesZone::IsInsideNoGrenadesZone(CBaseEntity *pOther)
+                                : vecHighAngVelocity;
+        break;
+
+    case SOLUTION_ONE:
+        vecLaunchVelocity = vecLowAngVelocity;
+        break;
+
+    case SOLUTION_TWO:
+        vecLaunchVelocity = vecHighAngVelocity;
+        break;
+
+    default:
+        break;
+    }
+
+    return vecLaunchVelocity;
+}
+void CTriggerMomentumCatapult::LaunchAtDirection(CBaseEntity *pOther)
+{
+    pOther->SetGroundEntity(nullptr);
+    Vector vecLaunchDir = vec3_origin;
+    AngleVectors(m_vLaunchDirection, &vecLaunchDir);
+    pOther->SetAbsVelocity(m_flPlayerSpeed * vecLaunchDir);
+    m_OnCatapulted.FireOutput(pOther, this);
+}
+
+void CTriggerMomentumCatapult::LaunchAtTarget(CBaseEntity *pOther)
+{
+    pOther->SetGroundEntity(nullptr);
+    if (m_iUseExactVelocity)
+    {
+        Vector vecLaunchVelocity = CalculateLaunchVelocityExact(pOther);
+        pOther->SetAbsVelocity(vecLaunchVelocity);      
+    }
+    else
+    {
+        Vector vecLaunchVelocity = CalculateLaunchVelocity(pOther);
+        pOther->SetAbsVelocity(vecLaunchVelocity);
+    }
+
+    m_OnCatapulted.FireOutput(pOther, this);
+}
+
+void CTriggerMomentumCatapult::Launch(CBaseEntity *pOther)
+{
+    bool bLaunch = true;
+
+    // Check threshold
+    if (m_bUseThresholdCheck)
+    {
+        Vector vecPlayerVelocity = pOther->GetAbsVelocity();
+        float flPlayerSpeed = vecPlayerVelocity.Length();
+        bLaunch = false;
+
+        // From VDC
+        if (flPlayerSpeed > m_flPlayerSpeed - (m_flPlayerSpeed * m_flLowerThreshold) &&
+            flPlayerSpeed < m_flPlayerSpeed + (m_flPlayerSpeed * m_flUpperThreshold))
+        {
+            if (m_bUseLaunchTarget)
+            {
+                Vector vecAbsDifference = m_hLaunchTarget->GetAbsOrigin() - pOther->GetAbsOrigin();
+
+                float flPlayerEntryAng = DotProduct(vecAbsDifference.Normalized(), vecPlayerVelocity.Normalized());
+
+                // VDC uses brackets so inclusive??
+                if (flPlayerEntryAng >= m_flEntryAngleTolerance)
+                {
+                    if (m_bOnlyCheckVelocity)
+                    {
+                        m_OnCatapulted.FireOutput(pOther, this);
+                        return;
+                    }
+                    bLaunch = true;
+                }
+            }
+            else
+            {
+                Vector vecLaunchDir = vec3_origin;
+                AngleVectors(m_vLaunchDirection, &vecLaunchDir);
+                float flPlayerEntryAng = DotProduct(vecLaunchDir.Normalized(), vecPlayerVelocity.Normalized());
+
+                // VDC uses brackets so inclusive??
+                if (flPlayerEntryAng >= m_flEntryAngleTolerance)
+                {
+                    if (m_bOnlyCheckVelocity)
+                    {
+                        m_OnCatapulted.FireOutput(pOther, this);
+                        return;
+                    }
+                    bLaunch = true;
+                }
+            }
```

Comment:
These two branches of code use very similar logic. Consider pulling the common code out:
```suggestion
        float flPlayerEntryAng;
        if (m_bUseLaunchTarget)
        {
            Vector vecAbsDifference = m_hLaunchTarget->GetAbsOrigin() - pOther->GetAbsOrigin();
            flPlayerEntryAng = DotProduct(vecAbsDifference.Normalized(), vecPlayerVelocity.Normalized());
        }
        else
        {
             Vector vecLaunchDir = vec3_origin;
             AngleVectors(m_vLaunchDirection, &vecLaunchDir);
             flPlayerEntryAng = DotProduct(vecLaunchDir.Normalized(), vecPlayerVelocity.Normalized());
         }

         // VDC uses brackets so inclusive??
         if (flPlayerEntryAng >= m_flEntryAngleTolerance)
         {
             if (m_bOnlyCheckVelocity)
             {
                 m_OnCatapulted.FireOutput(pOther, this);
                 return;
             }
             
             bLaunch = true;
         }
```